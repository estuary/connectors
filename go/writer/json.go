package writer

import (
	"compress/flate"
	"fmt"
	"io"

	"github.com/estuary/connectors/go/encrow"
	"github.com/klauspost/pgzip"
	"github.com/segmentio/encoding/json"
)

const (
	// JSON generally compresses well at minimum compression levels. Higher levels of compression
	// will usually take a lot more CPU while not providing much space savings.
	jsonCompressionlevel = flate.BestSpeed

	// Snowflake docs (https://docs.snowflake.com/en/sql-reference/sql/put#usage-notes) recommend
	// data file sizes be in the range of 100-250MB for compressed data. Redshift docs
	// (https://docs.aws.amazon.com/redshift/latest/dg/t_splitting-data-files.html) suggest files
	// for compressed data be in the range of 1MB to 1GB, and all be about the same size. Bigquery
	// docs don't mention anything other than the files must be less than 4GB. So, a 250MB file size
	// seems reasonable to use across the board given the current materializations that use this.
	DefaultJsonFileSizeLimit = 250 * 1024 * 1024
)

type jsonConfig struct {
	disableCompression bool
	skipNulls          bool
}

type JsonOption func(*jsonConfig)

func WithJsonDisableCompression() JsonOption {
	return func(cfg *jsonConfig) {
		cfg.disableCompression = true
	}
}

func WithJsonSkipNulls() JsonOption {
	return func(cfg *jsonConfig) {
		cfg.skipNulls = true
	}
}

type JsonWriter struct {
	w     io.Writer // will be set to `gz` for compressed writes or `cwc` if compression is disabled
	cwc   *countingWriteCloser
	gz    *pgzip.Writer
	shape *encrow.Shape
	buf   []byte
}

// NewJsonWriter creates a JsonWriter from w. w is closed when JsonWriter is closed. If `fields`
// is nil, values will be encoded as a JSON array rather than as an object.
func NewJsonWriter(w io.WriteCloser, fields []string, opts ...JsonOption) *JsonWriter {
	var cfg jsonConfig
	for _, o := range opts {
		o(&cfg)
	}

	jw := &JsonWriter{
		cwc: &countingWriteCloser{w: w},
	}

	if !cfg.disableCompression {
		gz, err := pgzip.NewWriterLevel(jw.cwc, jsonCompressionlevel)
		if err != nil {
			// Only possible if compressionLevel is not valid.
			panic("invalid compression level for gzip.NewWriterLevel")
		}
		jw.gz = gz
		jw.w = gz
	} else {
		jw.w = jw.cwc
	}

	if fields != nil {
		jw.shape = encrow.NewShape(fields)
		// Setting TrustRawMessage here prevents unnecessary validation of pre-serialized JSON
		// received from the runtime, which we can assume to be valid (flow_document for example).
		// Note that we are also not setting SortMapKeys or EscapeHTML: Sorting keys is not needed
		// because encrow.Shape already sorts the top-level keys and any object values are already
		// serialized as JSON, and escaping HTML is not desired so as to avoid escaping values like
		// <, >, &, etc. if they are present in the materialized collection's data.
		jw.shape.SetFlags(json.TrustRawMessage)
		jw.shape.SetSkipNulls(cfg.skipNulls)
	}

	return jw
}

func (w *JsonWriter) Write(vals []any) (err error) {
	w.buf = w.buf[:0]
	if w.shape == nil {
		// Serialize as a JSON array of values.
		w.buf = append(w.buf, '[')
		for idx, v := range vals {
			if w.buf, err = json.Append(w.buf, v, json.TrustRawMessage); err != nil {
				return fmt.Errorf("encoding JSON array value: %w", err)
			}
			if idx != len(vals)-1 {
				w.buf = append(w.buf, ',')
			}
		}
		w.buf = append(w.buf, ']')
	} else {
		// Serialize as a JSON object.
		if w.buf, err = w.shape.Encode(w.buf, vals); err != nil {
			return fmt.Errorf("encoding shape: %w", err)
		}
	}

	w.buf = append(w.buf, '\n')

	if _, err := w.w.Write(w.buf); err != nil {
		return fmt.Errorf("writing gzip bytes: %w", err)
	}

	return nil
}

func (w *JsonWriter) Written() int {
	return w.cwc.written
}

// Close closes the underlying gzip writer if compression is enabled, flushing its data and writing
// the GZIP footer. It also closes the underlying io.WriteCloser that was used to initialize the
// counting writer.
func (w *JsonWriter) Close() error {
	if w.gz != nil {
		if err := w.gz.Close(); err != nil {
			return fmt.Errorf("closing gzip writer: %w", err)
		}
	}

	if err := w.cwc.Close(); err != nil {
		return fmt.Errorf("closing counting writer: %w", err)
	}
	return nil
}
