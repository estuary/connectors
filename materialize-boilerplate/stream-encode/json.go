package stream_encode

import (
	"compress/flate"
	"fmt"
	"io"

	"github.com/estuary/connectors/go/encrow"
	"github.com/klauspost/compress/gzip"
	"github.com/segmentio/encoding/json"
)

const (
	// JSON generally compresses well at minimum compression levels. Higher levels of compression
	// will usually take a lot more CPU while not providing much space savings.
	jsonCompressionlevel = flate.BestSpeed

	// Snowflake docs (https://docs.snowflake.com/en/sql-reference/sql/put#usage-notes) recommend
	// data file sizes be in the range of 100-250MB for compressed data. Redshift docs
	// (https://docs.aws.amazon.com/redshift/latest/dg/t_splitting-data-files.html) suggest files
	// for compressed data be in the range of 1MB to 1GB, and all be about the same size. Bigquery
	// docs don't mention anything other than the files must be less than 4GB. So, a 250MB file size
	// seems reasonable to use across the board given the current materializations that use this.
	DefaultJsonFileSizeLimit = 250 * 1024 * 1024
)

type jsonConfig struct {
	disableCompression bool
}

type JsonOption func(*jsonConfig)

func WithJsonDisableCompression() JsonOption {
	return func(cfg *jsonConfig) {
		cfg.disableCompression = true
	}
}

type JsonEncoder struct {
	w     io.Writer // will be set to `gz` for compressed writes or `cwc` if compression is disabled
	cwc   *countingWriteCloser
	gz    *gzip.Writer
	shape *encrow.Shape
	buf   []byte
}

// NewJsonEncoder creates a JsonEncoder from w. w is closed when JsonEncoder is closed. If `fields`
// is nil, values will be encoded as a JSON array rather than as an object.
func NewJsonEncoder(w io.WriteCloser, fields []string, opts ...JsonOption) *JsonEncoder {
	var cfg jsonConfig
	for _, o := range opts {
		o(&cfg)
	}

	enc := &JsonEncoder{
		cwc: &countingWriteCloser{w: w},
	}

	if !cfg.disableCompression {
		gz, err := gzip.NewWriterLevel(enc.cwc, jsonCompressionlevel)
		if err != nil {
			// Only possible if compressionLevel is not valid.
			panic("invalid compression level for gzip.NewWriterLevel")
		}
		enc.gz = gz
		enc.w = gz
	} else {
		enc.w = enc.cwc
	}

	if fields != nil {
		enc.shape = encrow.NewShape(fields)
		// Setting TrustRawMessage here prevents unnecessary validation of pre-serialized JSON
		// received from the runtime, which we can assume to be valid (flow_document for example).
		// Note that we are also not setting SortMapKeys or EscapeHTML: Sorting keys is not needed
		// because encrow.Shape already sorts the top-level keys and any object values are already
		// serialized as JSON, and escaping HTML is not desired so as to avoid escaping values like
		// <, >, &, etc. if they are present in the materialized collection's data.
		enc.shape.SetFlags(json.TrustRawMessage)
	}

	return enc
}

func (e *JsonEncoder) Encode(vals []any) (err error) {
	if e.shape == nil {
		// Serialize as a JSON array of values.
		e.buf = e.buf[:0]
		e.buf = append(e.buf, '[')
		for idx, v := range vals {
			if e.buf, err = json.Append(e.buf, v, json.TrustRawMessage); err != nil {
				return fmt.Errorf("encoding JSON array value: %w", err)
			}
			if idx != len(vals)-1 {
				e.buf = append(e.buf, ',')
			}
		}
		e.buf = append(e.buf, ']')
	} else {
		// Serialize as a JSON object.
		if e.buf, err = e.shape.Encode(e.buf, vals); err != nil {
			return fmt.Errorf("encoding shape: %w", err)
		}
	}

	e.buf = append(e.buf, '\n')

	if _, err := e.w.Write(e.buf); err != nil {
		return fmt.Errorf("writing gzip bytes: %w", err)
	}

	return nil
}

func (e *JsonEncoder) Written() int {
	return e.cwc.written
}

// Close closes the underlying gzip writer if compression is enabled, flushing its data and writing
// the GZIP footer. It also closes the underlying io.WriteCloser that was used to initialize the
// counting encoder.
func (e *JsonEncoder) Close() error {
	if e.gz != nil {
		if err := e.gz.Close(); err != nil {
			return fmt.Errorf("closing gzip writer: %w", err)
		}
	}

	if err := e.cwc.Close(); err != nil {
		return fmt.Errorf("closing counting writer: %w", err)
	}
	return nil
}
