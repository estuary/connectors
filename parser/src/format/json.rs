//! Parser for the json format. This will accept any stream of JSON values separated by whitespace.
//! It allows any amount of whitespace (including newlines) within and in between records.
use super::{Input, Output, ParseError, Parser};
use crate::ParseConfig;

struct JsonParser;

pub fn new_parser() -> Box<dyn Parser> {
    Box::new(JsonParser)
}

impl Parser for JsonParser {
    fn parse(&self, _config: &ParseConfig, content: Input) -> Result<Output, ParseError> {
        let deser = serde_json::de::Deserializer::from_reader(content).into_iter();
        let wrapped = deser.map(|r| r.map_err(Into::into));
        Ok(Box::new(wrapped))
    }
}
