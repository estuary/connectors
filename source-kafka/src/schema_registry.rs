use anyhow::Result;
use futures::stream::{self, StreamExt};
use reqwest::Client;
use serde::Deserialize;
use std::collections::{HashMap, HashSet};

const TOPIC_KEY_SUFFIX: &str = "-key";
const TOPIC_VALUE_SUFFIX: &str = "-value";
const CONCURRENT_SCHEMA_REQUESTS: usize = 10;

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
struct FetchedSchema {
    #[serde(default = "SchemaType::default")]
    schema_type: SchemaType,
    schema: String,
    reference: Option<SchemaReference>,
}

#[derive(Deserialize, Debug)]
struct SchemaReference {
    name: String,
    subject: String,
    version: u32,
}

#[derive(Deserialize, Debug)]
struct FetchedLatestVersion {
    id: u32,
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "UPPERCASE")]
enum SchemaType {
    Avro,
    Json,
    Protobuf,
}

impl SchemaType {
    fn default() -> Self {
        SchemaType::Avro
    }
}

#[derive(Debug)]
pub enum RegisteredSchema {
    Avro(apache_avro::Schema),
    Json(serde_json::Value),
    Protobuf, // TODO(whb): Protobuf support.
}

#[derive(Debug, Default)]
pub struct TopicSchema {
    pub key: Option<RegisteredSchema>,
    pub value: Option<RegisteredSchema>,
}

pub struct SchemaRegistryClient {
    endpoint: String,
    http: Client,
    username: String,
    password: String,
}

impl SchemaRegistryClient {
    pub fn new(endpoint: String, username: String, password: String) -> SchemaRegistryClient {
        SchemaRegistryClient {
            endpoint: endpoint.to_string(),
            http: reqwest::Client::default(),
            username,
            password,
        }
    }

    pub async fn schemas_for_topics(
        &self,
        topics: &[String],
    ) -> Result<HashMap<String, TopicSchema>> {
        let applicable_topics: HashSet<String> = topics.iter().cloned().collect();

        let subjects: Vec<String> = self
            .http
            .get(format!("{}/subjects", self.endpoint))
            .basic_auth(&self.username, Some(&self.password))
            .send()
            .await?
            .json()
            .await?;

        let filter_by_suffix = |s: &str, suffix: &str| {
            if let Some(s) = s.strip_suffix(suffix) {
                if !applicable_topics.contains(s) {
                    return None;
                }
                return Some(s.to_string());
            }
            None
        };

        let topics_with_key_schema: HashSet<String> = subjects
            .iter()
            .filter_map(|s| filter_by_suffix(s, TOPIC_KEY_SUFFIX))
            .collect();

        let topics_with_value_schema: HashSet<String> = subjects
            .iter()
            .filter_map(|s| filter_by_suffix(s, TOPIC_VALUE_SUFFIX))
            .collect();

        let schema_futures: Vec<_> = applicable_topics
            .iter()
            .filter_map(|topic| {
                let need_key = topics_with_key_schema.contains(topic);
                let need_value = topics_with_value_schema.contains(topic);
                if !need_key && !need_value {
                    return None;
                }
                Some(async move {
                    let mut schema = TopicSchema {
                        key: None,
                        value: None,
                    };

                    if need_key {
                        schema.key = Some(self.fetch_latest_schema(topic, true).await?)
                    }
                    if need_value {
                        schema.value = Some(self.fetch_latest_schema(topic, false).await?)
                    }

                    Ok::<(String, TopicSchema), anyhow::Error>((topic.to_owned(), schema))
                })
            })
            .collect();

        stream::iter(schema_futures)
            .buffer_unordered(CONCURRENT_SCHEMA_REQUESTS)
            .collect::<Vec<_>>()
            .await
            .into_iter()
            .collect::<Result<HashMap<String, TopicSchema>>>()
    }

    pub async fn fetch_schema(&self, id: u32) -> Result<RegisteredSchema> {
        let fetched: FetchedSchema = self
            .http
            .get(format!("{}/schemas/ids/{}", self.endpoint, id))
            .basic_auth(&self.username, Some(&self.password))
            .send()
            .await?
            .json()
            .await?;

        match fetched.schema_type {
            SchemaType::Avro => {
                // TODO(whb): Resolve references.
                let schema = apache_avro::Schema::parse_str(&fetched.schema)
                    .expect("failed to parse avro schema");
                Ok(RegisteredSchema::Avro(schema))
            }
            SchemaType::Json => {
                // TODO(whb): Resolve references for JSON too.
                let schema =
                    serde_json::from_str(&fetched.schema).expect("failed to parse json schema");
                Ok(RegisteredSchema::Json(schema))
            }
            SchemaType::Protobuf => Ok(RegisteredSchema::Protobuf),
        }
    }

    async fn fetch_latest_version(&self, subject: &str) -> Result<u32> {
        let fetched: FetchedLatestVersion = self
            .http
            .get(format!(
                "{}/subjects/{}/versions/latest",
                self.endpoint, subject
            ))
            .basic_auth(&self.username, Some(&self.password))
            .send()
            .await?
            .json()
            .await?;
        Ok(fetched.id)
    }

    async fn fetch_latest_schema(&self, topic: &str, key: bool) -> Result<RegisteredSchema> {
        let subject = format!(
            "{}{}",
            topic,
            if key {
                TOPIC_KEY_SUFFIX
            } else {
                TOPIC_VALUE_SUFFIX
            }
        );
        let version = self.fetch_latest_version(subject.as_str()).await?;
        self.fetch_schema(version).await
    }
}
