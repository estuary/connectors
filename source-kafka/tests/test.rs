use std::time::Duration;

use anyhow::Result;
use apache_avro::types::{Record, Value};
use apache_avro::Schema;
use rdkafka::admin::{AdminClient, AdminOptions, NewTopic, TopicReplication};
use rdkafka::message::{Header, OwnedHeaders, ToBytes};
use rdkafka::producer::{FutureProducer, FutureRecord};
use rdkafka::ClientConfig;
use schema_registry_converter::async_impl::avro::AvroEncoder;
use schema_registry_converter::async_impl::json::JsonEncoder;
use schema_registry_converter::async_impl::schema_registry::SrSettings;
use schema_registry_converter::schema_registry_common::SubjectNameStrategy;
use serde_json::json;

#[test]
fn test_spec() {
    let output = std::process::Command::new("flowctl")
        .args(["raw", "spec", "--source", "tests/test.flow.yaml"])
        .output()
        .unwrap();

    assert!(output.status.success());
    let got: serde_json::Value = serde_json::from_slice(&output.stdout).unwrap();
    insta::assert_snapshot!(serde_json::to_string_pretty(&got).unwrap());
}

#[tokio::test]
async fn test_discover() {
    setup_test().await;

    let output = std::process::Command::new("flowctl")
        .args([
            "raw",
            "discover",
            "--source",
            "tests/test.flow.yaml",
            "-o",
            "json",
            "--emit-raw",
        ])
        .output()
        .unwrap();

    assert!(output.status.success());

    let snap = std::str::from_utf8(&output.stdout)
        .unwrap()
        .lines()
        .map(|line| serde_json::from_str::<serde_json::Value>(line).unwrap())
        .map(|line| serde_json::to_string_pretty(&line).unwrap())
        .reduce(|snap, line| format!("{}\n{}", snap, line))
        .unwrap();

    insta::assert_snapshot!(snap);
}

#[tokio::test]
async fn test_capture() {
    setup_test().await;

    let output = std::process::Command::new("flowctl")
        .args([
            "preview",
            "--source",
            "tests/test.flow.yaml",
            "--sessions",
            "1",
            "--delay",
            "2s",
            "--output-state",
        ])
        .output()
        .unwrap();

    println!("{}", std::str::from_utf8(&output.stderr).unwrap());

    assert!(output.status.success());

    let snap = std::str::from_utf8(&output.stdout).unwrap();

    insta::assert_snapshot!(snap);
}

#[tokio::test]
async fn test_capture_resume() {
    setup_test().await;

    let initial_state = json!({
      "bindingStateV1": {
        "avro-topic": {
          "partitions": {
            "1": 1,
            "2": 1
          }
        },
        "json-schema-topic": {
          "partitions": {
            "1": 1,
            "2": 1
          }
        },
        "json-raw-topic": {
          "partitions": {
            "1": 1,
            "2": 1
          }
        }
      }
    });

    let output = std::process::Command::new("flowctl")
        .args([
            "preview",
            "--source",
            "tests/test.flow.yaml",
            "--sessions",
            "1",
            "--delay",
            "2s",
            "--output-state",
            "--initial-state",
            &initial_state.to_string(),
        ])
        .output()
        .unwrap();

    assert!(output.status.success());

    let snap = std::str::from_utf8(&output.stdout).unwrap();

    insta::assert_snapshot!(snap);
}

async fn setup_test() {
    let bootstrap_servers = "localhost:9092";
    let schema_registry_endpoint = "http://localhost:8081";
    let num_messages = 9;
    let num_partitions = 3;
    let topic_replication = 1;

    let test_cases: &[(&dyn TestDataEncoder, &str)] = &[
        (&AvroTestDataEncoder::new(), "avro-topic"),
        (&JsonSchemaTestDataEncoder::new(), "json-schema-topic"),
        (&JsonRawTestDataEncoder::new(), "json-raw-topic"),
    ];

    let http = reqwest::Client::default();

    let admin: AdminClient<_> = ClientConfig::new()
        .set("bootstrap.servers", bootstrap_servers)
        .create()
        .unwrap();

    let producer: FutureProducer = ClientConfig::new()
        .set("bootstrap.servers", bootstrap_servers)
        .create()
        .unwrap();

    let opts = AdminOptions::default().request_timeout(Some(Duration::from_secs(1)));

    for (enc, topic) in test_cases {
        admin.delete_topics(&[topic], &opts).await.unwrap();
        admin
            .create_topics(
                &[NewTopic::new(
                    topic,
                    num_partitions,
                    TopicReplication::Fixed(topic_replication),
                )],
                &opts,
            )
            .await
            .unwrap()
            .into_iter()
            .collect::<Result<Vec<_>, _>>()
            .unwrap();

        // Registry schemas if the encoder uses schemas.
        if let Some(schema_type) = enc.schema_type_string() {
            for (topic, suffix, schema) in [
                (topic, "key", &enc.key_schema_string()),
                (topic, "value", &enc.payload_schema_string()),
            ] {
                // Try to delete the existing schema if it exists, ignoring any errors
                // that are returned, which may be 404's.
                http.delete(format!(
                    "{}/subjects/{}-{}",
                    schema_registry_endpoint, topic, suffix
                ))
                .send()
                .await
                .unwrap();

                // You have to do a "soft" delete before a permanent "hard" delete.
                http.delete(format!(
                    "{}/subjects/{}-{}?permanent=true",
                    schema_registry_endpoint, topic, suffix
                ))
                .send()
                .await
                .unwrap();

                // Register the schema, which must be successful.
                assert!(http
                    .post(format!(
                        "{}/subjects/{}-{}/versions",
                        schema_registry_endpoint, topic, suffix
                    ))
                    .json(&json!({"schema": schema, "schemaType": schema_type}))
                    .send()
                    .await
                    .unwrap()
                    .status()
                    .is_success());
            }
        }

        // Populate regular "data" records.
        for idx in 0..num_messages {
            send_message(
                topic,
                &enc.key_for_idx(idx, topic).await,
                Some(&enc.payload_for_idx(idx, topic).await),
                idx,
                num_partitions,
                &producer,
            )
            .await;
        }

        // Populate deletion records.
        for idx in 0..num_partitions {
            send_message(
                topic,
                &enc.key_for_idx(idx as usize, topic).await,
                None::<&[u8]>,
                idx as usize,
                num_partitions,
                &producer,
            )
            .await;
        }
    }
}

async fn send_message<K, P>(
    topic: &str,
    key: &K,
    payload: Option<&P>,
    idx: usize,
    num_partitions: i32,
    producer: &FutureProducer,
) where
    K: ToBytes + ?Sized,
    P: ToBytes + ?Sized,
{
    let mut rec = FutureRecord::to(topic)
        .partition(idx as i32 % num_partitions)
        .key(key)
        .timestamp(unix_millis_fixture(idx))
        .headers(OwnedHeaders::new().insert(Header {
            key: "header-key",
            value: Some(&format!("header-value-{}", idx)),
        }));

    if let Some(payload) = payload {
        rec = rec.payload(payload);
    }

    producer.send(rec, None).await.unwrap();
}

fn unix_millis_fixture(idx: usize) -> i64 {
    ((idx + 1) * 86_400_000) as i64
}

#[async_trait::async_trait]
trait TestDataEncoder {
    async fn key_for_idx<'a>(&'a self, idx: usize, topic: &'a str) -> Vec<u8>;
    async fn payload_for_idx<'a>(&'a self, idx: usize, topic: &'a str) -> Vec<u8>;
    fn key_schema_string(&self) -> String;
    fn payload_schema_string(&self) -> String;
    fn schema_type_string(&self) -> Option<String>;
}

struct AvroTestDataEncoder {}

impl AvroTestDataEncoder {
    fn new() -> Self {
        AvroTestDataEncoder {}
    }
}

#[async_trait::async_trait]
impl TestDataEncoder for AvroTestDataEncoder {
    async fn key_for_idx<'a>(&'a self, idx: usize, topic: &'a str) -> Vec<u8> {
        let enc = AvroEncoder::new(SrSettings::new(String::from("http://localhost:8081")));
        let schema =
            Schema::parse(&serde_json::from_str(&self.key_schema_string()).unwrap()).unwrap();

        let mut key = Record::new(&schema).unwrap();
        key.put("idx", Value::Int(idx as i32));
        key.put(
            "nested",
            Value::Record(vec![("sub_id".to_string(), Value::Int(idx as i32))]),
        );

        enc.encode_value(
            key.into(),
            &SubjectNameStrategy::TopicNameStrategy(topic.to_string(), true),
        )
        .await
        .unwrap()
    }

    async fn payload_for_idx<'a>(&'a self, idx: usize, topic: &'a str) -> Vec<u8> {
        let enc = AvroEncoder::new(SrSettings::new(String::from("http://localhost:8081")));
        let schema =
            Schema::parse(&serde_json::from_str(&self.payload_schema_string()).unwrap()).unwrap();

        let mut value = Record::new(&schema).unwrap();
        value.put("value", Value::String(format!("value-{}", idx)));

        enc.encode_value(
            value.into(),
            &SubjectNameStrategy::TopicNameStrategy(topic.to_string(), false),
        )
        .await
        .unwrap()
    }

    fn key_schema_string(&self) -> String {
        let parsed = Schema::parse(&json!({
          "type": "record",
          "name": "AvroKey",
          "fields": [
            {
              "name": "idx",
              "type": "int"
            },
            {
              "name": "nested",
              "type": {
                "type": "record",
                "name": "NestedAvroKeyRecord",
                "fields": [
                  {
                    "name": "sub_id",
                    "type": "int"
                  }
                ]
              }
            }
          ]
        }))
        .unwrap();

        parsed.canonical_form()
    }

    fn payload_schema_string(&self) -> String {
        let parsed = Schema::parse(&json!({
          "type": "record",
          "name": "AvroValue",
          "fields": [
            {
              "name": "value",
              "type": "string"
            }
          ]
        }))
        .unwrap();

        parsed.canonical_form()
    }

    fn schema_type_string(&self) -> Option<String> {
        Some("AVRO".to_string())
    }
}

struct JsonSchemaTestDataEncoder {}

impl JsonSchemaTestDataEncoder {
    fn new() -> Self {
        Self {}
    }
}

#[async_trait::async_trait]
impl TestDataEncoder for JsonSchemaTestDataEncoder {
    async fn key_for_idx<'a>(&'a self, idx: usize, topic: &'a str) -> Vec<u8> {
        let enc = JsonEncoder::new(SrSettings::new(String::from("http://localhost:8081")));
        enc.encode(
            &json!({
                "idx": idx,
                "nested": {
                    "sub_id": idx
                },
            }),
            SubjectNameStrategy::TopicNameStrategy(topic.to_string(), true),
        )
        .await
        .unwrap()
    }

    async fn payload_for_idx<'a>(&'a self, idx: usize, topic: &'a str) -> Vec<u8> {
        let enc = JsonEncoder::new(SrSettings::new(String::from("http://localhost:8081")));
        enc.encode(
            &json!({
                "value": format!("value-{}", idx),
            }),
            SubjectNameStrategy::TopicNameStrategy(topic.to_string(), false),
        )
        .await
        .unwrap()
    }

    fn key_schema_string(&self) -> String {
        serde_json::to_string(&json!({
            "$schema": "http://json-schema.org/draft-07/schema#",
            "title": "JsonKey",
            "type": "object",
            "properties": {
              "idx": {
                "type": "integer"
              },
              "nested": {
                "type": "object",
                "title": "NestedJsonKeyRecord",
                "properties": {
                  "sub_id": {
                    "type": "integer"
                  }
                },
                "required": ["sub_id"]
              }
            },
            "required": ["idx", "nested"],
            "additionalProperties": false
        }))
        .unwrap()
    }

    fn payload_schema_string(&self) -> String {
        serde_json::to_string(&json!({
          "$schema": "http://json-schema.org/draft-07/schema#",
          "title": "JsonValue",
          "type": "object",
          "properties": {
            "value": {
              "type": "string"
            }
          },
          "required": ["value"],
          "additionalProperties": false
        }))
        .unwrap()
    }

    fn schema_type_string(&self) -> Option<String> {
        Some("JSON".to_string())
    }
}

struct JsonRawTestDataEncoder {}

impl JsonRawTestDataEncoder {
    fn new() -> Self {
        Self {}
    }
}

#[async_trait::async_trait]
impl TestDataEncoder for JsonRawTestDataEncoder {
    async fn key_for_idx<'a>(&'a self, idx: usize, _: &'a str) -> Vec<u8> {
        serde_json::to_vec(&json!({
            "key": idx,
        }))
        .unwrap()
    }

    async fn payload_for_idx<'a>(&'a self, idx: usize, _: &'a str) -> Vec<u8> {
        serde_json::to_vec(&json!({
            "payload": idx,
        }))
        .unwrap()
    }

    fn key_schema_string(&self) -> String {
        panic!("not implemented")
    }

    fn payload_schema_string(&self) -> String {
        panic!("not implemented")
    }

    fn schema_type_string(&self) -> Option<String> {
        None
    }
}
