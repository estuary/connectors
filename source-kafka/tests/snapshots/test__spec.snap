---
source: tests/test.rs
expression: "serde_json::to_string_pretty(&got).unwrap()"
---
{
  "configSchema": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "properties": {
      "bootstrap_servers": {
        "description": "The initial servers in the Kafka cluster to initially connect to, separated by commas. The Kafka client will be informed of the rest of the cluster nodes by connecting to one of these nodes.",
        "order": 0,
        "title": "Bootstrap Servers",
        "type": "string"
      },
      "credentials": {
        "description": "The connection details for authenticating a client connection to Kafka via SASL. When not provided, the client connection will attempt to use PLAINTEXT (insecure) protocol. This must only be used in dev/test environments.",
        "discriminator": {
          "propertyName": "auth_type"
        },
        "oneOf": [
          {
            "properties": {
              "auth_type": {
                "const": "user_password",
                "default": "user_password",
                "order": 0,
                "type": "string"
              },
              "mechanism": {
                "default": "PLAIN",
                "description": "The SASL Mechanism describes how to exchange and authenticate clients/servers.",
                "enum": [
                  "PLAIN",
                  "SCRAM-SHA-256",
                  "SCRAM-SHA-512"
                ],
                "order": 1,
                "title": "SASL Mechanism",
                "type": "string"
              },
              "password": {
                "order": 3,
                "secret": true,
                "title": "Password",
                "type": "string"
              },
              "username": {
                "order": 2,
                "secret": true,
                "title": "Username",
                "type": "string"
              }
            },
            "required": [
              "auth_type",
              "mechanism",
              "password",
              "username"
            ],
            "title": "SASL (User & Password)"
          },
          {
            "properties": {
              "auth_type": {
                "const": "AWS",
                "default": "AWS",
                "type": "string"
              },
              "aws_access_key_id": {
                "order": 0,
                "title": "AWS Access Key ID",
                "type": "string"
              },
              "aws_secret_access_key": {
                "order": 1,
                "secret": true,
                "title": "AWS Secret Access Key",
                "type": "string"
              },
              "region": {
                "order": 2,
                "title": "AWS Region",
                "type": "string"
              }
            },
            "required": [
              "auth_type",
              "aws_access_key_id",
              "aws_secret_access_key",
              "region"
            ],
            "title": "AWS MSK IAM"
          }
        ],
        "order": 1,
        "title": "Credentials",
        "type": "object"
      },
      "schema_registry": {
        "description": "Connection details for interacting with a schema registry. This is necessary for processing messages encoded with Avro.",
        "order": 3,
        "properties": {
          "endpoint": {
            "description": "Schema registry API endpoint. For example: https://registry-id.us-east-2.aws.confluent.cloud",
            "order": 0,
            "title": "Schema Registry Endpoint",
            "type": "string"
          },
          "password": {
            "description": "Schema registry password to use for authentication. If you are using Confluent Cloud, this will be the 'Secret' from your schema registry API key.",
            "order": 2,
            "secret": true,
            "title": "Schema Registry Password",
            "type": "string"
          },
          "username": {
            "description": "Schema registry username to use for authentication. If you are using Confluent Cloud, this will be the 'Key' from your schema registry API key.",
            "order": 1,
            "title": "Schema Registry Username",
            "type": "string"
          }
        },
        "required": [
          "endpoint",
          "password",
          "username"
        ],
        "title": "Schema Registry",
        "type": "object"
      },
      "tls": {
        "default": "system_certificates",
        "description": "Controls how should TLS certificates be found or used.",
        "enum": [
          "system_certificates"
        ],
        "order": 2,
        "title": "TLS Settings",
        "type": "string"
      }
    },
    "required": [
      "bootstrap_servers",
      "credentials"
    ],
    "title": "Kafka Source Configuration",
    "type": "object"
  },
  "documentationUrl": "https://go.estuary.dev/source-kafka",
  "protocol": 3032023,
  "resourceConfigSchema": {
    "$schema": "https://json-schema.org/draft/2019-09/schema",
    "properties": {
      "topic": {
        "description": "Kafka topic to capture messages from.",
        "title": "Topic",
        "type": "string"
      }
    },
    "required": [
      "topic"
    ],
    "title": "Resource",
    "type": "object"
  },
  "resourcePathPointers": [
    "/topic"
  ]
}
